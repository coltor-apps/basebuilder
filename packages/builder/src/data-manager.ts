import {
  createSubscriptionManager,
  type Subscribe,
} from "./subscription-manager";

export function createDataManager<TData>(initialData: TData): {
  getData: () => TData;
  setData: (data: TData) => TData;
  subscribe: Subscribe<TData>;
} {
  let data: TData = initialData;

  function getData(): TData {
    return data;
  }

  const { notify, subscribe } = createSubscriptionManager<TData>();

  return {
    subscribe,
    getData,
    setData(newData) {
      data = newData;

      notify(data);

      return data;
    },
  };
}
