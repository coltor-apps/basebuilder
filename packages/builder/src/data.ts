import { createSubscriptionManager, type Subscribe } from "./subscription";

export function createDataManager<TData>(data: TData): {
  getData: () => TData;
  setData: (setter: (oldData: TData) => TData) => TData;
  subscribe: Subscribe<TData>;
} {
  let localData: TData = data;

  function getData(): TData {
    return localData;
  }

  const { notify, subscribe } = createSubscriptionManager<TData>();

  return {
    subscribe,
    getData,
    setData(setter) {
      localData = setter(getData());

      notify(localData);

      return localData;
    },
  };
}
