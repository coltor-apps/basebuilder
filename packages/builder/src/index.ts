export { createBuilder } from "./builder";
export type { Builder, BuilderEntities } from "./builder";

export { createEntity } from "./entity";
export type { Entity } from "./entity";

export { createInput } from "./input";
export type { Input } from "./input";

export { SchemaValidationError as EntityValidationError, createStore } from "./store";
export type {
  Store,
  StoreData,
  StoreEntity,
  SchemaValidationErrorCode as EntityValidationErrorCause,
} from "./store";

export { getEntitiesNamesExcept } from "./utils";
