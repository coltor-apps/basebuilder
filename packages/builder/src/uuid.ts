import crypto from "crypto";

export function generateUuid(): string {
  return crypto.randomUUID();
}

export function validateUuid(id: string): void {
  if (
    typeof id !== "string" ||
    !/^(?:[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}|00000000-0000-0000-0000-000000000000)$/i.test(
      id,
    )
  ) {
    throw new Error(`The entity id '${id}' is invalid.`);
  }
}
