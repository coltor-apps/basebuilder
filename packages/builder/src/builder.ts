import { type Entity } from "./entity";
import { type Input } from "./input";
import { generateUuid, validateUuid } from "./uuid";

export type BuilderEntities = ReadonlyArray<
  Entity<string, ReadonlyArray<Input<string, unknown>>, unknown>
>;

type ChildrenAllowed<TEntities extends BuilderEntities> = {
  [K in TEntities[number]["name"]]?:
    | ReadonlyArray<TEntities[number]["name"]>
    | true;
};

export type Builder<
  TEntities extends BuilderEntities = BuilderEntities,
  TChildrenAllowed extends ChildrenAllowed<TEntities> = Record<
    string,
    true | ReadonlyArray<string> | undefined
  >,
  TParentRequired extends ReadonlyArray<
    TEntities[number]["name"]
  > = ReadonlyArray<string>,
> = {
  entities: TEntities;
  entityId: {
    generate: () => string;
    validate: (id: string) => void;
  };
  childrenAllowed: TChildrenAllowed;
  parentRequired: TParentRequired;
};

type OptionalBuilderArgs = "childrenAllowed" | "parentRequired";

export function createBuilder<
  const TEntities extends BuilderEntities,
  const TChildrenAllowed extends ChildrenAllowed<TEntities> = Record<
    string,
    never
  >,
  const TParentRequired extends ReadonlyArray<TEntities[number]["name"]> = [],
>(
  options: Omit<
    Builder<TEntities, TChildrenAllowed, TParentRequired>,
    OptionalBuilderArgs | "entityId"
  > &
    Partial<
      Pick<
        Builder<TEntities, TChildrenAllowed, TParentRequired>,
        OptionalBuilderArgs
      >
    > & {
      entityId?: Partial<
        Builder<TEntities, TChildrenAllowed, TParentRequired>["entityId"]
      >;
    },
): Builder<TEntities, TChildrenAllowed, TParentRequired> {
  return {
    ...options,
    entityId: {
      generate: options.entityId?.generate ?? generateUuid,
      validate: options.entityId?.validate ?? validateUuid,
    },
    childrenAllowed: options.childrenAllowed ?? ({} as TChildrenAllowed),
    parentRequired:
      options.parentRequired ??
      ([] as ReadonlyArray<unknown> as TParentRequired),
  };
}

export function getEntityDefinition(
  entityType: string,
  builder: Builder,
): Builder["entities"][number] | undefined {
  return builder.entities.find(
    (builderEntity) => builderEntity.name === entityType,
  );
}

export function ensureEntityIsRegistered(
  entityType: string,
  builder: Builder,
): Builder["entities"][number] {
  const entityDefinition = getEntityDefinition(entityType, builder);

  if (!entityDefinition) {
    throw new Error(`Unkown entity type "${entityType}".`);
  }

  return entityDefinition;
}

export function ensureEntityInputIsRegistered(
  entityType: string,
  inputName: string,
  builder: Builder,
): Input {
  const entityDefinition = ensureEntityIsRegistered(entityType, builder);

  const input = entityDefinition.inputs.find(
    (input) => input.name === inputName,
  );

  if (!input) {
    throw new Error(`Unkown entity input "${inputName}".`);
  }

  return input;
}

export function ensureEntityInputsAreRegistered(
  entityType: string,
  inputNames: Array<string>,
  builder: Builder,
): Array<Input> {
  const inputs = inputNames.map((inputName) =>
    ensureEntityInputIsRegistered(entityType, inputName, builder),
  );

  return inputs;
}

export function isEntityChildAllowed(
  entityType: string,
  childEntityType: string,
  builder: Builder,
): boolean {
  const allowedChildren = builder.childrenAllowed[entityType];

  if (!allowedChildren) {
    return false;
  }

  return allowedChildren === true || allowedChildren.includes(childEntityType);
}

export function isEntityParentRequired(
  entityType: string,
  builder: Builder,
): boolean {
  return builder.parentRequired.includes(entityType);
}

export function ensureEntityChildAllowed(
  entityType: string,
  childEntityType: string,
  builder: Builder,
): void {
  if (!isEntityChildAllowed(entityType, childEntityType, builder)) {
    throw new Error("Child is not allowed.");
  }
}

export function ensureEntityCanLackParent(
  entityType: string,
  builder: Builder,
): void {
  if (isEntityParentRequired(entityType, builder)) {
    throw new Error("A parent is required.");
  }
}
