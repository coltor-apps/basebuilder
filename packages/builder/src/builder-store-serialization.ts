import { type Builder } from "./builder";
import {
  type BuilderStoreData,
  type BuilderStoreEntity,
  type SerializedBuilderStoreData,
} from "./builder-store";

export function serializeBuilderStoreData<TBuilder extends Builder>(
  data: BuilderStoreData<TBuilder>,
): SerializedBuilderStoreData<TBuilder> {
  const newEntities: SerializedBuilderStoreData<TBuilder>["schema"]["entities"] =
    {};

  for (const [id, entity] of data.schema.entities) {
    const { children, ...entityData } = entity;

    newEntities[id] = {
      ...entityData,
      ...(children ? { children: Array.from(children) } : {}),
      inputs:
        entityData.inputs as unknown as SerializedBuilderStoreData<TBuilder>["schema"]["entities"][string]["inputs"],
    };
  }

  return {
    schema: {
      root: Array.from(data.schema.root),
      entities: newEntities,
    },
    entitiesInputsErrors: Object.fromEntries(data.entitiesInputsErrors),
  };
}

export function deserializeBuilderStoreData<TBuilder extends Builder>(
  data: SerializedBuilderStoreData<TBuilder>,
): BuilderStoreData<TBuilder> {
  return {
    schema: {
      entities: new Map(
        Object.entries(data.schema.entities).map(([id, entity]) => [
          id,
          {
            ...entity,
            ...(entity.children ? { children: new Set(entity.children) } : {}),
            inputs:
              entity.inputs as unknown as BuilderStoreEntity<TBuilder>["inputs"],
          },
        ]),
      ),
      root: new Set(data.schema.root),
    },
    entitiesInputsErrors: new Map(Object.entries(data.entitiesInputsErrors)),
  };
}
