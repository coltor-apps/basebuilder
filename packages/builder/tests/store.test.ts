import { afterEach, describe, expect, it, vi } from "vitest";

import { createBuilder, createEntity, createInput, createStore } from "../src";
import * as schemaExports from "../src/schema";
import * as uuidExports from "../src/uuid";

describe("store", () => {
  afterEach(() => {
    vi.restoreAllMocks();
  });

  it("can be created without a schema", () => {
    const builder = createBuilder({
      entities: [],
    });

    const store = createStore(builder);

    expect(store).toMatchInlineSnapshot(`
      {
        "addEntity": [Function],
        "builder": {
          "childrenAllowed": {},
          "entities": [],
          "entityId": {
            "generate": [Function],
            "validate": [Function],
          },
          "parentRequired": [],
        },
        "getData": [Function],
        "getSchema": [Function],
        "subscribe": [Function],
      }
    `);
  });

  it("can be created with a non-empty schema", () => {
    const builder = createBuilder({
      entities: [
        createEntity({
          name: "text",
          inputs: [
            createInput({
              name: "label",
              validate(value) {
                return value;
              },
            }),
          ],
        }),
      ],
    });

    const store = createStore(builder, {
      entities: {
        "6e0035c3-0d4c-445f-a42b-2d971225447c": {
          type: "text",
          inputs: {
            label: "test",
          },
        },
      },
      root: ["6e0035c3-0d4c-445f-a42b-2d971225447c"],
    });

    expect(store).toMatchInlineSnapshot(`
      {
        "addEntity": [Function],
        "builder": {
          "childrenAllowed": {},
          "entities": [
            {
              "defaultValue": [Function],
              "inputs": [
                {
                  "defaultValue": [Function],
                  "name": "label",
                  "validate": [Function],
                },
              ],
              "isValueAllowed": false,
              "name": "text",
              "shouldBeProcessed": [Function],
              "validate": [Function],
            },
          ],
          "entityId": {
            "generate": [Function],
            "validate": [Function],
          },
          "parentRequired": [],
        },
        "getData": [Function],
        "getSchema": [Function],
        "subscribe": [Function],
      }
    `);
  });

  it("can be created with an empty schema", () => {
    const builder = createBuilder({
      entities: [],
    });

    const store = createStore(builder, { entities: {}, root: [] });

    expect(store).toMatchInlineSnapshot(`
      {
        "addEntity": [Function],
        "builder": {
          "childrenAllowed": {},
          "entities": [],
          "entityId": {
            "generate": [Function],
            "validate": [Function],
          },
          "parentRequired": [],
        },
        "getData": [Function],
        "getSchema": [Function],
        "subscribe": [Function],
      }
    `);
  });

  it("can retrieve the data", () => {
    const builder = createBuilder({
      entities: [
        createEntity({
          name: "text",
          inputs: [
            createInput({
              name: "label",
              validate(value) {
                return value;
              },
            }),
          ],
        }),
      ],
    });

    const store = createStore(builder, {
      entities: {
        "6e0035c3-0d4c-445f-a42b-2d971225447c": {
          type: "text",
          inputs: {
            label: "test",
          },
        },
      },
      root: ["6e0035c3-0d4c-445f-a42b-2d971225447c"],
    });

    expect(store.getData()).toMatchInlineSnapshot(`
      {
        "entities": Map {
          "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
            "inputs": {
              "label": "test",
            },
            "type": "text",
          },
        },
        "root": Set {
          "6e0035c3-0d4c-445f-a42b-2d971225447c",
        },
      }
    `);
  });

  it("can return the schema", () => {
    const builder = createBuilder({
      entities: [
        createEntity({
          name: "text",
          inputs: [
            createInput({
              name: "label",
              validate(value) {
                return value;
              },
            }),
          ],
        }),
      ],
    });

    const store = createStore(builder, {
      entities: {
        "6e0035c3-0d4c-445f-a42b-2d971225447c": {
          type: "text",
          inputs: {
            label: "test",
          },
        },
      },
      root: ["6e0035c3-0d4c-445f-a42b-2d971225447c"],
    });

    expect(store.getSchema()).toMatchInlineSnapshot(`
      {
        "entities": {
          "6e0035c3-0d4c-445f-a42b-2d971225447c": {
            "inputs": {
              "label": "test",
            },
            "type": "text",
          },
        },
        "root": [
          "6e0035c3-0d4c-445f-a42b-2d971225447c",
        ],
      }
    `);
  });

  it("can add entities and notify listeners", () => {
    vi.spyOn(uuidExports, "generateUuid").mockImplementation(
      () => "6e0035c3-0d4c-445f-a42b-2d971225447c",
    );

    const builder = createBuilder({
      entities: [
        createEntity({
          name: "text",
          inputs: [
            createInput({
              name: "label",
              validate(value) {
                return value;
              },
            }),
          ],
        }),
      ],
    });

    const store = createStore(builder, {
      entities: {},
      root: [],
    });

    const listener = vi.fn();

    store.subscribe(listener);

    store.addEntity({ type: "text", inputs: {} });

    expect(store.getData()).toMatchInlineSnapshot(`
      {
        "entities": Map {
          "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
            "inputs": {},
            "type": "text",
          },
        },
        "root": Set {
          "6e0035c3-0d4c-445f-a42b-2d971225447c",
        },
      }
    `);

    expect(listener).toHaveBeenCalledWith({
      entities: new Map([
        ["6e0035c3-0d4c-445f-a42b-2d971225447c", { inputs: {}, type: "text" }],
      ]),
      root: new Set(["6e0035c3-0d4c-445f-a42b-2d971225447c"]),
    });
  });

  it("validates the parent when adding entities", () => {
    const ensureEntityParentIdHasValidReferenceMock = vi.spyOn(
      schemaExports,
      "ensureEntityParentIdHasValidReference",
    );

    const ensureEntityChildAllowedMock = vi.spyOn(
      schemaExports,
      "ensureEntityChildAllowed",
    );

    const ensureEntityCanLackParentMock = vi.spyOn(
      schemaExports,
      "ensureEntityCanLackParent",
    );

    vi.spyOn(uuidExports, "generateUuid").mockImplementation(
      () => "6e0035c3-0d4c-445f-a42b-2d971225447c",
    );

    const builder = createBuilder({
      entities: [
        createEntity({
          name: "test",
        }),
      ],
      childrenAllowed: {
        test: true,
      },
    });

    const store = createStore(builder, {
      entities: {
        "c1ab14a4-41db-4531-9a58-4825a9ef6d26": {
          type: "test",
          inputs: {},
        },
      },
      root: [],
    });

    store.addEntity({
      type: "test",
      inputs: {},
      parentId: "c1ab14a4-41db-4531-9a58-4825a9ef6d26",
    });

    expect(ensureEntityParentIdHasValidReferenceMock).toHaveBeenCalledWith(
      {
        id: "6e0035c3-0d4c-445f-a42b-2d971225447c",
        inputs: {},
        parentId: "c1ab14a4-41db-4531-9a58-4825a9ef6d26",
        type: "test",
      },
      {
        "c1ab14a4-41db-4531-9a58-4825a9ef6d26": {
          type: "test",
          inputs: {},
        },
      },
    );

    expect(ensureEntityChildAllowedMock).toHaveBeenCalledWith(
      {
        id: "c1ab14a4-41db-4531-9a58-4825a9ef6d26",
        type: "test",
        inputs: {},
      },
      {
        id: "6e0035c3-0d4c-445f-a42b-2d971225447c",
        inputs: {},
        parentId: "c1ab14a4-41db-4531-9a58-4825a9ef6d26",
        type: "test",
      },
      builder,
    );

    expect(ensureEntityCanLackParentMock).not.toHaveBeenCalled();
  });

  it("ensures an entity can lack a parent ID when added", () => {
    const ensureEntityCanLackParentMock = vi.spyOn(
      schemaExports,
      "ensureEntityCanLackParent",
    );

    vi.spyOn(uuidExports, "generateUuid").mockImplementation(
      () => "6e0035c3-0d4c-445f-a42b-2d971225447c",
    );

    const builder = createBuilder({
      entities: [
        createEntity({
          name: "test",
        }),
      ],
    });

    const store = createStore(builder, {
      entities: {},
      root: [],
    });

    store.addEntity({
      type: "test",
      inputs: {},
    });

    expect(ensureEntityCanLackParentMock).toHaveBeenCalledWith(
      {
        id: "6e0035c3-0d4c-445f-a42b-2d971225447c",
        inputs: {},
        type: "test",
      },
      builder,
    );
  });
});
