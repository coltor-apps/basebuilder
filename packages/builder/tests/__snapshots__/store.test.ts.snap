// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`store > can add entities to a parent entity 1`] = `
{
  "entities": Map {
    "51324b32-adc3-4d17-a90e-66b5453935bd" => {
      "children": Set {
        "6e0035c3-0d4c-445f-a42b-2d971225447c",
      },
      "inputs": {},
      "type": "test",
    },
    "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
      "inputs": {},
      "parentId": "51324b32-adc3-4d17-a90e-66b5453935bd",
      "type": "test",
    },
  },
  "root": Set {
    "51324b32-adc3-4d17-a90e-66b5453935bd",
  },
}
`;

exports[`store > can add entities to a parent entity at a specific index 1`] = `
{
  "entities": Map {
    "51324b32-adc3-4d17-a90e-66b5453935bd" => {
      "children": Set {
        "6e0035c3-0d4c-445f-a42b-2d971225447c",
        "a02cd91c-d982-4e80-8fa4-184e9fe2b0b5",
      },
      "inputs": {},
      "type": "test",
    },
    "a02cd91c-d982-4e80-8fa4-184e9fe2b0b5" => {
      "inputs": {},
      "parentId": "51324b32-adc3-4d17-a90e-66b5453935bd",
      "type": "test",
    },
    "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
      "inputs": {},
      "parentId": "51324b32-adc3-4d17-a90e-66b5453935bd",
      "type": "test",
    },
  },
  "root": Set {
    "51324b32-adc3-4d17-a90e-66b5453935bd",
  },
}
`;

exports[`store > can add entities to root 1`] = `
{
  "entities": Map {
    "51324b32-adc3-4d17-a90e-66b5453935bd" => {
      "inputs": {},
      "type": "test",
    },
    "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
      "inputs": {},
      "type": "test",
    },
  },
  "root": Set {
    "51324b32-adc3-4d17-a90e-66b5453935bd",
    "6e0035c3-0d4c-445f-a42b-2d971225447c",
  },
}
`;

exports[`store > can add entities to root at specific index 1`] = `
{
  "entities": Map {
    "51324b32-adc3-4d17-a90e-66b5453935bd" => {
      "inputs": {},
      "type": "test",
    },
    "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
      "inputs": {},
      "type": "test",
    },
  },
  "root": Set {
    "6e0035c3-0d4c-445f-a42b-2d971225447c",
    "51324b32-adc3-4d17-a90e-66b5453935bd",
  },
}
`;

exports[`store > can be created with a non-empty schema 1`] = `
[MockFunction baseValidateSchema] {
  "calls": [
    [
      {
        "childrenAllowed": {},
        "entities": [
          {
            "defaultValue": [Function],
            "inputs": [
              {
                "defaultValue": [Function],
                "name": "label",
                "validate": [Function],
              },
            ],
            "isValueAllowed": false,
            "name": "text",
            "shouldBeProcessed": [Function],
            "validate": [Function],
          },
        ],
        "entityId": {
          "generate": [Function],
          "validate": [Function],
        },
        "parentRequired": [],
      },
      {
        "entities": {
          "6e0035c3-0d4c-445f-a42b-2d971225447c": {
            "inputs": {
              "label": "test",
            },
            "type": "text",
          },
        },
        "root": [
          "6e0035c3-0d4c-445f-a42b-2d971225447c",
        ],
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": {
        "entities": {
          "6e0035c3-0d4c-445f-a42b-2d971225447c": {
            "inputs": {
              "label": "test",
            },
            "type": "text",
          },
        },
        "root": [
          "6e0035c3-0d4c-445f-a42b-2d971225447c",
        ],
      },
    },
  ],
}
`;

exports[`store > can be created with a non-empty schema 2`] = `
{
  "addEntity": [Function],
  "builder": {
    "childrenAllowed": {},
    "entities": [
      {
        "defaultValue": [Function],
        "inputs": [
          {
            "defaultValue": [Function],
            "name": "label",
            "validate": [Function],
          },
        ],
        "isValueAllowed": false,
        "name": "text",
        "shouldBeProcessed": [Function],
        "validate": [Function],
      },
    ],
    "entityId": {
      "generate": [Function],
      "validate": [Function],
    },
    "parentRequired": [],
  },
  "deleteEntity": [Function],
  "getData": [Function],
  "getSchema": [Function],
  "subscribe": [Function],
  "updateEntity": [Function],
}
`;

exports[`store > can be created with an empty schema 1`] = `
[MockFunction baseValidateSchema] {
  "calls": [
    [
      {
        "childrenAllowed": {},
        "entities": [],
        "entityId": {
          "generate": [Function],
          "validate": [Function],
        },
        "parentRequired": [],
      },
      {
        "entities": {},
        "root": [],
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": {
        "entities": {},
        "root": [],
      },
    },
  ],
}
`;

exports[`store > can be created with an empty schema 2`] = `
{
  "addEntity": [Function],
  "builder": {
    "childrenAllowed": {},
    "entities": [],
    "entityId": {
      "generate": [Function],
      "validate": [Function],
    },
    "parentRequired": [],
  },
  "deleteEntity": [Function],
  "getData": [Function],
  "getSchema": [Function],
  "subscribe": [Function],
  "updateEntity": [Function],
}
`;

exports[`store > can be created without a schema 1`] = `
{
  "addEntity": [Function],
  "builder": {
    "childrenAllowed": {},
    "entities": [],
    "entityId": {
      "generate": [Function],
      "validate": [Function],
    },
    "parentRequired": [],
  },
  "deleteEntity": [Function],
  "getData": [Function],
  "getSchema": [Function],
  "subscribe": [Function],
  "updateEntity": [Function],
}
`;

exports[`store > can delete entities and cascade delete their children 1`] = `
{
  "entities": Map {
    "49e91328-02bc-4daa-ab56-619554e85cff" => {
      "children": Set {},
      "inputs": {},
      "type": "test",
    },
  },
  "root": Set {
    "49e91328-02bc-4daa-ab56-619554e85cff",
  },
}
`;

exports[`store > can move an entity from root to a parent entity 1`] = `
{
  "entities": Map {
    "51324b32-adc3-4d17-a90e-66b5453935bd" => {
      "children": Set {
        "a02cd91c-d982-4e80-8fa4-184e9fe2b0b5",
        "6e0035c3-0d4c-445f-a42b-2d971225447c",
      },
      "inputs": {},
      "type": "test",
    },
    "a02cd91c-d982-4e80-8fa4-184e9fe2b0b5" => {
      "inputs": {},
      "parentId": "51324b32-adc3-4d17-a90e-66b5453935bd",
      "type": "test",
    },
    "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
      "inputs": {},
      "parentId": "51324b32-adc3-4d17-a90e-66b5453935bd",
      "type": "test",
    },
  },
  "root": Set {
    "51324b32-adc3-4d17-a90e-66b5453935bd",
  },
}
`;

exports[`store > can move an entity from root to a parent entity at a specific index 1`] = `
{
  "entities": Map {
    "51324b32-adc3-4d17-a90e-66b5453935bd" => {
      "children": Set {
        "6e0035c3-0d4c-445f-a42b-2d971225447c",
        "a02cd91c-d982-4e80-8fa4-184e9fe2b0b5",
      },
      "inputs": {},
      "type": "test",
    },
    "a02cd91c-d982-4e80-8fa4-184e9fe2b0b5" => {
      "inputs": {},
      "parentId": "51324b32-adc3-4d17-a90e-66b5453935bd",
      "type": "test",
    },
    "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
      "inputs": {},
      "parentId": "51324b32-adc3-4d17-a90e-66b5453935bd",
      "type": "test",
    },
  },
  "root": Set {
    "51324b32-adc3-4d17-a90e-66b5453935bd",
  },
}
`;

exports[`store > can move an entity in a parent entity 1`] = `
{
  "entities": Map {
    "51324b32-adc3-4d17-a90e-66b5453935bd" => {
      "children": Set {
        "6e0035c3-0d4c-445f-a42b-2d971225447c",
        "a02cd91c-d982-4e80-8fa4-184e9fe2b0b5",
      },
      "inputs": {},
      "type": "test",
    },
    "a02cd91c-d982-4e80-8fa4-184e9fe2b0b5" => {
      "inputs": {},
      "parentId": "51324b32-adc3-4d17-a90e-66b5453935bd",
      "type": "test",
    },
    "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
      "inputs": {},
      "parentId": "51324b32-adc3-4d17-a90e-66b5453935bd",
      "type": "test",
    },
  },
  "root": Set {
    "51324b32-adc3-4d17-a90e-66b5453935bd",
  },
}
`;

exports[`store > can move an entity in root 1`] = `
{
  "entities": Map {
    "51324b32-adc3-4d17-a90e-66b5453935bd" => {
      "inputs": {},
      "type": "test",
    },
    "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
      "inputs": {},
      "type": "test",
    },
  },
  "root": Set {
    "6e0035c3-0d4c-445f-a42b-2d971225447c",
    "51324b32-adc3-4d17-a90e-66b5453935bd",
  },
}
`;

exports[`store > can move an entity in root 2`] = `
{
  "entities": Map {
    "51324b32-adc3-4d17-a90e-66b5453935bd" => {
      "inputs": {},
      "type": "test",
    },
    "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
      "inputs": {},
      "type": "test",
    },
  },
  "root": Set {
    "51324b32-adc3-4d17-a90e-66b5453935bd",
    "6e0035c3-0d4c-445f-a42b-2d971225447c",
  },
}
`;

exports[`store > can move an entity in root 3`] = `
{
  "entities": Map {
    "51324b32-adc3-4d17-a90e-66b5453935bd" => {
      "inputs": {},
      "type": "test",
    },
    "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
      "inputs": {},
      "type": "test",
    },
  },
  "root": Set {
    "51324b32-adc3-4d17-a90e-66b5453935bd",
    "6e0035c3-0d4c-445f-a42b-2d971225447c",
  },
}
`;

exports[`store > can move an entity to root 1`] = `
{
  "entities": Map {
    "51324b32-adc3-4d17-a90e-66b5453935bd" => {
      "children": Set {},
      "inputs": {},
      "type": "test",
    },
    "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
      "inputs": {},
      "type": "test",
    },
  },
  "root": Set {
    "51324b32-adc3-4d17-a90e-66b5453935bd",
    "6e0035c3-0d4c-445f-a42b-2d971225447c",
  },
}
`;

exports[`store > can move an entity to root at a specific index 1`] = `
{
  "entities": Map {
    "51324b32-adc3-4d17-a90e-66b5453935bd" => {
      "children": Set {},
      "inputs": {},
      "type": "test",
    },
    "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
      "inputs": {},
      "type": "test",
    },
  },
  "root": Set {
    "6e0035c3-0d4c-445f-a42b-2d971225447c",
    "51324b32-adc3-4d17-a90e-66b5453935bd",
  },
}
`;

exports[`store > can retrieve the data 1`] = `
{
  "entities": Map {
    "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
      "inputs": {
        "label": "test",
      },
      "type": "text",
    },
  },
  "root": Set {
    "6e0035c3-0d4c-445f-a42b-2d971225447c",
  },
}
`;

exports[`store > can return the schema 1`] = `
{
  "entities": {
    "6e0035c3-0d4c-445f-a42b-2d971225447c": {
      "inputs": {
        "label": "test",
      },
      "type": "text",
    },
  },
  "root": [
    "6e0035c3-0d4c-445f-a42b-2d971225447c",
  ],
}
`;

exports[`store > doesn't update the entity when no mutation fields were provided 1`] = `
{
  "entities": Map {
    "6e0035c3-0d4c-445f-a42b-2d971225447c" => {
      "inputs": {},
      "type": "test",
    },
  },
  "root": Set {
    "6e0035c3-0d4c-445f-a42b-2d971225447c",
  },
}
`;

exports[`store > throws when trying to add an entity to a non existent parent entity 1`] = `"Entity with ID \\"invalid\\" was not found."`;

exports[`store > throws when trying to delete non existent entity 1`] = `"Entity with ID \\"test\\" was not found."`;

exports[`store > throws when trying to move an entity to a non existent parent entity 1`] = `"Entity with ID \\"invalid\\" was not found."`;

exports[`store > throws when trying to update a non existent entity 1`] = `"Entity with ID \\"invalid\\" was not found."`;
